unified_address.py

```diff
@@ -51,46 +52,35 @@ def decode_unified(addr_str):
     suffix = decoded[-16:]
     # check trailing padding bytes
     assert suffix == padding(hrp)
-    decoded = decoded[:-16]
+    rest = decoded[:-16]
 
-    s = 0
-    acc = []
     result = {}
-    for b in decoded:
-        if s == 0:
-            receiver_type = b
-            s = 1
-        elif s == 1:
-            receiver_len = b
-            expected_len = {0: 20, 1: 20, 2: 43, 3: 43}.get(receiver_type)
-            if expected_len is not None:
-                assert receiver_len == expected_len, "incorrect receiver length"
-            s = 2
-        elif s == 2:
-            if len(acc) < receiver_len:
-                acc.append(b)
-            
-            if len(acc) == receiver_len:
-                if receiver_type == 0 or receiver_type == 1:
-                    assert not ('transparent' in result), "duplicate transparent receiver detected"
-                    assert len(acc) == 20
-                    result['transparent'] = bytes(acc)
-                    acc = []
-                    s = 0
-
-                elif receiver_type == 2:
-                    assert not ('sapling' in result), "duplicate sapling receiver detected"
-                    assert len(acc) == 43
-                    result['sapling'] = bytes(acc)
-                    acc = []
-                    s = 0
-
-                elif receiver_type == 3:
-                    assert not ('orchard' in result), "duplicate orchard receiver detected"
-                    assert len(acc) == 43
-                    result['orchard'] = bytes(acc)
-                    acc = []
-                    s = 0
+    while len(rest) > 0:
+        (receiver_type, rest) = parse_compact_size(rest)
+        (receiver_len, rest) = parse_compact_size(rest)
```

Comment:
We should be enforcing the maximum CompactSize length here (by passing the appropriate flag, or having `parse_compact_size` default to enforcing it).